Handle repeated predicate pushdown into Hive connector

martint opened the pull request:
The previous implementation was only considering the first
attempt where a filter is pushed down into the Hive connector.
As a result, for a query like this, the partition filter above
the bottommost filter would be ignored:

    SELECT * FROM (
        SELECT * FROM t WHERE a in (1, 2)
    ) u
    WHERE u.pk = 'b';

martint committed:
Handle repeated predicate pushdown into Hive connector

The previous implementation was only considering the first
attempt where a filter is pushed down into the Hive connector.
As a result, for a query like this, the partition filter above
the bottommost filter would be ignored:

    SELECT * FROM (
        SELECT * FROM t WHERE a in (1, 2)
    ) u
    WHERE u.pk = 'b';

findepi commented:
@martint make sure this change doesn't increase the  number of times we query the Metastore for partitions.
Each call may be costly (in seconds or sometimes even worse).
You can easily track this with https://github.com/prestosql/presto/pull/946

Also, make sure that queries like
```
SELECT * FROM (
    SELECT * FROM t WHERE pk > 0
) u
WHERE u.pk = 'b';
```
do not hit the `hive.max-partitions-per-scan` limit.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1691,10 +1691,6 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         checkArgument(!handle.getAnalyzePartitionValues().isPresent() || constraint.getSummary().isAll(), "Analyze should not have a constraint");
 
-        if (handle.getPartitions().isPresent()) {
-            return Optional.empty(); // TODO: optimize multiple calls to applyFilter
```
Alternative (or complementary) approach would be to "squash" all the filters before invoking `applyFilter` for the first time.

martint replied:
Yes, although that's tricky. It'd require running PredicatePushdown before all other rules run, but PPD benefits from other things such as unaliasSymbols, etc, and all the other rules might benefit from the result of applyFilter...

I was thinking we may need a different approach, but haven't settled on what, exactly. Some options:
1. Skip trying to enumerate partitions in the connector if there are no predicates over partition keys. This could limit the ability of predicate inference to use that info for propagating to the other side of a join, though, especially when there are only a few partitions.
2. Try enumerating partitions up to a point. If there are more than a certain threshold, stop and say "ALL" (and record that fact to avoid doing it all over again unless a different filter is provided). This still can require calling the metastore multiple times if there multiple filters stack on each other, but at least can avoid enumerating all partitions.

findepi replied:
>  especially when there are only a few partitions. 

is this often the case?

> Try enumerating partitions up to a point. If there are more than a certain threshold, stop and say "ALL"

We've seen that in some cases, the call for partitions (even with _some_ filter) can take very long time (>20s).
The reason is that the call translates to metastore db query containing `partitonName like `partkey1=%/partkey2=x` (when only `partkey2 is provided).
Metastore db can decide that filter is not filtering enough (the constant prefix is too short to use an index) and go full-table-scan over partitions.  🕤

Of course, there are other APIs we could call and https://github.com/prestosql/presto/issues/611 is all about that. But it won't be solved tomorrow.

Now, in some cases we may need to repeat a call. Or, do we? Maybe if we got part list once, we can only fine-tune the selection in memory, without a trip to HMS.
Regardless, since HMS can in some cases have a big problem answering the call, it's important to derive as much information as possible from the query, before making first call.

> It'd require running PredicatePushdown before all other rules run, but PPD benefits from other things such as unaliasSymbols, etc, and all the other rules might benefit from the result of applyFilter...

That's true. Part list for one joined table can result in some additional conditions for the other table.
Anyway, we should derived as much information from the query itself before making first trip to HMS.
Think of it as a preparatory PPD (a bit like "dry-run"), then HMS, then PPD to distribute the information. (Then, maybe, refine part selection, this time without calling HMS.)

martint replied:
> Think of it as a preparatory PPD (a bit like "dry-run"), then HMS, then PPD to distribute the information. (Then, maybe, refine part selection, this time without calling HMS.)

Yes, that would be *somewhat* like option 1 above. Don't call the HMS until there's at least one filter over a partition key. But it's not exactly the same as what you describe because there are no "stages" for the optimization process. There's no way for the connector (or engine) to know that nothing else will be pushed down close to the TableScan.

As a workaround, for now, I'm looking into adding a call to PredicatePushdown above all the other optimizers. This should at least cause all the filters that can be squashed down in a single pass to be moved close to the table scan. We need to think about how this works in the long term, but #611 looks promising.

martint replied:
> As a workaround, for now, I'm looking into adding a call to PredicatePushdown above all the other optimizers

Ugh, and that doesn't work because PPD changes the plan enough to make decorrelation rules fail.

electrum replied:
> Now, in some cases we may need to repeat a call. Or, do we? Maybe if we got part list once, we can only fine-tune the selection in memory, without a trip to HMS.

Subsequent `applyFilter` calls only restrict the filter, never widen it, so the initial list of partitions can only become smaller (or unchanged).

findepi replied:
> Subsequent `applyFilter` calls only restrict the filter, never widen it, so the initial list of partitions can only become smaller (or unchanged).

Right. But we still need to ensure we gather as many filters as possible to avoid overloading coordinator (and HMS) when doing this initial listing (a.k.a. hitting `hive.max-partitions-per-scan` limit).

sopel39 replied:
> Ugh, and that doesn't work because PPD changes the plan enough to make decorrelation rules fail.

@martin Could you give an example of that? It seems that PPD could provide better filters so that decorrelation could be even more efficient. For example, pushing equality conditions to lateral join.

martint committed:
Return ALL for enforced contraint

NONE means no rows would be produced, which is incorrect.

martint committed:
Delay pushdown into connector

Execute the rules after all the predicate pushdowns, decorrelations and
other simplifications have executed to avoid calling into the connectors
multiple times if possible (these actions can be expensive for some connectors
such as Hive).

martint commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -2541,6 +2542,102 @@ public void testPredicatePushDownToTableScan()
+    public void testPartitionPruning()
+    {
+        assertUpdate("CREATE TABLE test_partition_pruning (v bigint, k varchar) WITH (partitioned_by = array['k'])");
+        assertUpdate("INSERT INTO test_partition_pruning (v, k) VALUES (1, 'a'), (2, 'b'), (3, 'c'), (4, 'e')", 4);
```
I need the type of the field to be different from the value that gets used in the query so that the planner inserts some implicit coercions. Otherwise, the plan ends up with the "right enough" shape to not trigger the problem. I can make it a CTAS with casts or explicit value constructors, but I'll also add a comment, since this is a non-obvious detail.

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -2541,6 +2542,113 @@ public void testPredicatePushDownToTableScan()
+
+            assertQuery(query, "VALUES (2, 'b'), (4, 'e')");
+            assertConstraints(
+                    "SELECT * FROM (" +
```
I think you meant to use `query` here

electrum commented:
Merged, thanks!

electrum commented:
Merged as 74b7a7952a0abd736b9ed7b86793d6c5621fd455
